backend/internetarchive/internetarchive.go

```diff
@@ -0,0 +1,1086 @@
+// relative to f.root
+func (f *Fs) split(rootRelativePath string) (bucketName, bucketPath string) {
+	bucketName, bucketPath = bucket.Split(path.Join(f.root, rootRelativePath))
+	return f.opt.Enc.FromStandardName(bucketName), f.opt.Enc.FromStandardPath(bucketPath)
+}
+
+// split returns bucket and bucketPath from the object
+func (o *Object) split() (bucket, bucketPath string) {
+	return o.fs.split(o.remote)
+}
+
+func (f *Fs) getHashes(ctx context.Context, src fs.ObjectInfo) map[hash.Type]string {
+	hashMap := map[hash.Type]string{}
+	for _, ty := range f.Hashes().Array() {
+		sum, err := src.Hash(ctx, ty)
+		if err != nil || sum == "" {
+			continue
+		}
+		hashMap[ty] = sum
+	}
+	return hashMap
+}
+
+func (f *Fs) requestMetadata(ctx context.Context, bucket string) (result MetadataResponse, err error) {
+	var resp *http.Response
+	// make a GET request to (frontend)/metadata/:item/
+	opts := rest.Opts{
+		Method: "GET",
+		Path:   path.Join("/metadata/", bucket),
+	}
+
+	err = f.pacer.Call(func() (bool, error) {
+		resp, err = f.front.CallJSON(ctx, &opts, nil, &result)
+		return f.shouldRetry(resp, err)
+	})
+
+	return result, err
+}
+
+// list up all files/directories without any filters
+func (f *Fs) listAllUnconstrained(ctx context.Context, bucket string) (entries fs.DirEntries, err error) {
+	result, err := f.requestMetadata(ctx, bucket)
+	if err != nil {
+		return nil, err
+	}
+
+	knownDirs := map[string]time.Time{
+		"": time.Unix(0, 0),
+	}
+	for _, file := range result.Files {
+		dir := strings.Trim(betterPathDir(file.Name), "/")
+		nameWithBucket := path.Join(bucket, file.Name)
+
+		mtimeTime := file.parseMtime()
+
+		// populate children directories
+		child := dir
+		for {
+			if _, ok := knownDirs[child]; ok {
+				break
+			}
+			// directory
+			d := fs.NewDir(f.opt.Enc.ToStandardPath(path.Join(bucket, child)), mtimeTime)
+			entries = append(entries, d)
+
+			knownDirs[child] = mtimeTime
+			child = strings.Trim(betterPathDir(child), "/")
+		}
+		if _, ok := knownDirs[betterPathDir(file.Name)]; !ok {
+			continue
+		}
+
+		size := parseSize(file.Size)
+
+		o := makeValidObject(f, f.opt.Enc.ToStandardPath(nameWithBucket), file, mtimeTime, size)
+		entries = append(entries, o)
+	}
+
+	return entries, nil
+}
+
+func (f *Fs) waitFileUpload(ctx context.Context, reqPath string, newHashes map[hash.Type]string, newSize int64) (ret *Object, err error) {
+	bucket, bucketPath := f.split(reqPath)
+
+	ret = &Object{
+		fs:      f,
+		remote:  trimPathPrefix(path.Join(bucket, bucketPath), f.root, f.opt.Enc),
+		modTime: time.Unix(0, 0),
+		size:    -1,
+	}
+
+	if !f.opt.WaitArchive {
+		// user doesn't want to poll, let's not
+		ret2, err := f.NewObject(ctx, reqPath)
+		if err == nil {
+			ret2, ok := ret2.(*Object)
+			if ok {
+				ret = ret2
+			}
+		}
+		return ret, nil
+	}
+
+	isFirstTime := true
+	existed := false
+	oldMtime := ""
+	oldCrc32 := ""
+	unreliablePassCount := 0
+	for {
```

Comment:
This looks like it could loop forever?

It should probably have a timeout. Maybe the value of the WaitArchive parameter should be a maximum  `fs.Duration` to wait?

Reply:
> It should probably have a timeout. Maybe the value of the WaitArchive parameter should be a maximum `fs.Duration` to wait?

That's a good idea.

Reply:
I tried to change it to accept duration, but it gives me the following error:

```
=== RUN   TestIntegration
    fstests.go:418: Using remote "TestIA:lesmi-rclone-test/"
    fstests.go:442: 
                Error Trace:    fstests.go:442
                                                        internetarchive_test.go:13
                Error:          Received unexpected error:
                                couldn't parse config item "wait_archive" = "1h0m0s" as time.Duration: parsing "1h0m0s" as time.Duration failed: expected newline
                Test:           TestIntegration
                Messages:       unexpected error: couldn't parse config item "wait_archive" = "1h0m0s" as time.Duration: parsing "1h0m0s" as time.Duration failed: expected newline
--- FAIL: TestIntegration (0.00s)
FAIL
FAIL    command-line-arguments  0.028s
FAIL
```

Config section:

```
		}, {
			Name: "wait_archive",
			Help: `Timeout for waiting the server's processing tasks (specifically archive and book_op) to finish.
Only enable if you need to be guaranteed to be reflected after write operations.
0 to disable waiting. No errors to be thrown in case of timeout.`,
			Default:  fs.Duration(0),
			Advanced: true,
		}, {
```

Entry in rclone.conf:

```
[TestIA]
type = internetarchive
access_key_id = [REDACTED]
secret_access_key = [REDACTED]
wait_archive = 1h0m0s
```

Reply:
I must have used `WaitArchive     fs.Duration          `config:"wait_archive"``. Resolved